tools/wpt/browser.py

```diff
@@ -1423,10 +1433,167 @@ class Safari(Browser):
     product = "safari"
     requirements = "requirements_safari.txt"
 
-    def download(self, dest=None, channel=None, rename=None):
-        raise NotImplementedError
+    def _find_downloads(self):
+        def text_content(e):
+            return etree.tostring(e, encoding="unicode", method="text")
+
+        self.logger.info("Finding STP download URLs")
+        resp = get("https://developer.apple.com/safari/download/")
+
+        doc = html5lib.parse(
+            resp.content,
+            "etree",
+            namespaceHTMLElements=False,
+            transport_encoding=resp.encoding,
+        )
+        ascii_ws = re.compile(r"[\x09\x0A\x0C\x0D\x20]+")
+
+        downloads = []
+        for candidate in doc.iterfind(".//li[@class]"):
+            class_names = set(ascii_ws.split(candidate.attrib["class"]))
+            if {"download", "dmg", "zip"} & class_names:
+                downloads.append(candidate)
+
+        stp_link_text = re.compile(r"^\s*Safari\s+Technology\s+Preview\s+for\s+macOS")
+        requirement = re.compile(
+            r"Requires\s+macOS\s+([0-9\.]+)\s+(?:or\s+later|beta)."
+        )
+
+        stp_downloads = []
+        for download in downloads:
+            for link in download.iterfind(".//a[@href]"):
+                if stp_link_text.match(text_content(link)):
+                    break
+            else:
+                continue
+
+            m = requirement.search(text_content(download))
+            if m:
+                version = m.group(1)
+
+                # This assumes the current macOS numbering, whereby X.Y is compatible
+                # with X.(Y+1), e.g. 12.4 is compatible with 12.3, but 13.0 isn't
+                # compatible with 12.3. This doesn't handle the former 10.* numbering.
+                if "." in version:
+                    spec = SpecifierSet(f"~={version}")
+                else:
+                    spec = SpecifierSet(f"=={version}.*")
+
+                stp_downloads.append((spec, link.attrib["href"]))
+
+        self.logger.info(
+            "Found STP URLs for macOS " + ", ".join(str(dl[0]) for dl in stp_downloads)
+        )
+        return stp_downloads
+
+    def _download_image(self, downloads, dest, system_version=None):
+        if system_version is None:
+            system_version, _, _ = platform.mac_ver()
+
+        for version_spec, url in downloads:
+            if system_version in version_spec:
+                self.logger.debug(f"Will download Safari for {version_spec}")
+                break
+        else:
+            raise ValueError(f"no download for {system_version}")
+
+        self.logger.info(f"Downloading Safari from {url}")
+        resp = get(url)
+
+        filename = None
+
+        content_disposition = resp.headers.get("content-disposition")
+        if content_disposition:
+            filenames = re.findall("filename=(.+)", content_disposition)
```

Comment:
This was copy/pasta from the Firefox code

Reply:
Well it still seems pretty broken :p Given that "extract the filename from a header" seems like it shouldn't be very specific to the browser, maybe extract this into a function rather than copy/paste.